### PR TITLE
Minor changes to fix some major issues on mobile when using gRPC.

### PR DESCRIPTION
**Pull request by slightfoot**
* Fixes bug when a ResponseStream is cancelled the underlying HTTP/2 stream is not disconnected so the transport never goes inactive/idle.
* ~~Add connectTimeout to socket connections.~~
* Fixes bug where connect state is mismanaged on socket error.

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @slightfoot isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

File: lib/src/client/connection.dart [hunk 1]
```diff
@@ -261,11 +270,11 @@ class ClientConnection {
     _connect();
   }
 
-  void _handleSocketClosed(_) {
+  void _handleSocketClosed() {
     _cancelTimer();
     _transport = null;
 
-    if (_state == ConnectionState.idle && _state == ConnectionState.shutdown) {
```
Comment (sigurdm) [inline lib/src/client/connection.dart]: Wow - that's nasty code!

**Comment by sigurdm** [comment]
And thanks a bunch for contributing

**Comment by slightfoot** [comment]
Sorry, I just don't have the time.

**Comment by sigurdm** [comment]
No problem - I will look into this

**Comment by magillus** [comment]
Please merge or fix this issue. I was trying to solve issue when http2 connection restarted (due to local gRPC server restart) and auto connection. 
I was going to prepare a PR then I saw this and applied that fix to my version - it works like a charm.

**Comment by slightfoot** [comment]
@sigurdm did you get around to the tests?

File: lib/src/client/options.dart [hunk 2]
```diff
@@ -21,11 +21,11 @@ import 'dart:math';
 import '../shared/security.dart';
 
 const defaultIdleTimeout = const Duration(minutes: 5);
+const defaultConnectTimeout = const Duration(seconds: 20);
```
Comment (sigurdm) [inline lib/src/client/options.dart]: I think we should not give a default connection timeout - that deviates from Socket.connect, and is therefore surprising.

File: lib/src/client/connection.dart [hunk 3]
```diff
@@ -103,7 +103,11 @@ class ClientConnection {
   Future<ClientTransportConnection> connectTransport() async {
     final securityContext = options.credentials.securityContext;
 
-    var socket = await Socket.connect(host, port);
+    var socket = await Socket.connect(
```
Comment (sigurdm) [inline lib/src/client/connection.dart]: We should probably surround this with a 
```
try {...} on SocketException catch {
// handle socket-error.
}
``

**Comment by sigurdm** [comment]
Sorry for the long delay.
I have tried to split this a bit - I created a pr with atest for the cancelling behavior.
I think we can should split this further into the connect timeout and the error handling. I think that would make it much easier to land this.

**Comment by VladimirLuzhin** [comment]
Hellow, is there any information when this pull request will be merge to the master branch?

**Comment by v0l** [comment]
This is badly needed..

**Comment by nesscx** [comment]
@sigurdm Will this be landed soon?

**Comment by njovy** [comment]
When is it going to be fixed? This is fairly critical for streams.

**Comment by VladimirLuzhin** [comment]
Can you tell me when this pull request will be uploaded?

**Comment by slightfoot** [comment]
@sigurdm do you want me split out the error and timeout behaviour?

**Comment by sigurdm** [comment]
Yes, that would be lovely

**Comment by slightfoot** [comment]
@sigurdm removed the connection timeout handling.

**Comment by slightfoot** [comment]
@sigurdm connection timeout is on separate branch.

https://github.com/slightfoot/grpc-dart/commit/11c9f4272484d339bdc840e8afef0e6cb49e735e

**Comment by slightfoot** [comment]
@sigurdm poke.. shall we try to get this merged before it goes out of date again.

**Comment by slightfoot** [comment]
@sigurdm I just had someone reach out to me because they were caught out by the issues this ticket fixes. Can we get this merged?

**Comment by mit-mit** [comment]
@mraleph can you help with merging this if we get this brought up to date?

**Comment by mraleph** [comment]
I would be happy to land this @slightfoot! However this requires rebase and (regression) tests which validate fixed issues.

**Comment by mit-mit** [comment]
@slightfoot any chance we can get you to rebase and add tests?

**Comment by mraleph** [comment]
Closing as assumed stale for now.

**Pull request closed without merging**